### PR TITLE
Fixing issues from Firebase

### PR DESCRIPTION
**Pull request by anamariamilcu**
Corrected errors generated by Firebase.
Correctly initialized the Firebase instance to be unique within the whole app using a singleton class to solve error with status code 500: Firebase App named AlphaBooks already exists. - caused by the fact that we were initializing the app multiple times.

### CODE REVIEW

File: src/main/java/com/google/sps/data/FirebaseSingletonApp.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+      // System.out.println(System.getProperty("user.dir"));
+      FileInputStream serviceAccount = new FileInputStream(System.getProperty("user.home") + 
+          "/step235-2020/serviceAccountKey.json");
+
+      FirebaseOptions options = new FirebaseOptions.Builder()
+          .setCredentials(GoogleCredentials.fromStream(serviceAccount))
+          .setDatabaseUrl("https://alphabooks-step-2020.firebaseio.com").build();
+
+      return FirebaseApp.initializeApp(options, "AlphaBooks");
+    } catch (FileNotFoundException e) {
```
Comment (alexcepoi) [inline src/main/java/com/google/sps/data/FirebaseSingletonApp.java]: Don't catch exceptions you don't know how to handle. Just let them through.

Yes, the logs are huge and it's pain to dig through them. But fighting java on this is pointless. Exceptions and long stack traces are simply something that you should get used to.

File: src/main/java/com/google/sps/data/FirebaseSingletonApp.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+  private static FirebaseApp instance = initialize();
+
+  /**
+   * Initialize firebase SDK and return the FirebaseApp object
+   */
+
+  private static FirebaseApp initialize() {
+    try {
+      // System.out.println(System.getProperty("user.dir"));
+      FileInputStream serviceAccount = new FileInputStream(System.getProperty("user.home") +
```
Comment (alexcepoi) [inline src/main/java/com/google/sps/data/FirebaseSingletonApp.java]: This won't work when you try to deploy it.

Just use the recommended way: https://firebase.google.com/docs/admin/setup#initialize-sdk

When running locally set the environment variable to the correct location. When running on appengine, you don't need to do anything.

File: src/main/java/com/google/sps/servlets/NewRequestServlet.java [hunk 3]
```diff
@@ -85,36 +84,37 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
     }
     
     // initialize firebase app 
-    FirebaseApp defaultApp = initializeFirebaseApp();
+    FirebaseApp firebaseApp = FirebaseSingletonApp.getInstance();
```
Comment (alexcepoi) [inline src/main/java/com/google/sps/servlets/NewRequestServlet.java]: Please write an unit test with a fake firestore to verify that your code is working after (and before) your changes.

https://firebase.google.com/docs/rules/unit-tests

File: src/main/java/com/google/sps/servlets/NewRequestServlet.java [hunk 4]
```diff
@@ -152,9 +141,11 @@ private void markUserAsCourier(String uid) {
...
-    if (userEntity != null && userEntity.getProperty("isCourier").equals("true")) {
-      // This is not the first request, so user is already marked as courier.
-      return;
+    if (userEntity != null) {
+      if(userEntity.getProperty("isCourier") != null && userEntity.getProperty("isCourier").equals("true")) {
+        // This is not the first request, so user is already marked as courier.
+        return;
+      }
     }
     if (userEntity == null) {
```
Comment (alexcepoi) [inline src/main/java/com/google/sps/servlets/NewRequestServlet.java]: else?

File: src/main/webapp/common.js [hunk 5]
```diff
@@ -104,15 +109,19 @@ function addMenu() {
   delivery_request.href = 'deliveryRequest.html';
   menuElement.append(delivery_request);
   const see_journeys = document.createElement('a');
-  let displayJourneys;
-  fetch('/user-data').then(response => response.json()).then((userInfo) => {
-    console.log(userInfo);
-    displayJourneys = userInfo.isCourier;
+  firebase.auth().onAuthStateChanged(function(user) {
```
Comment (alexcepoi) [inline src/main/webapp/common.js:120]: This javascript code should have tests I think.

Otherwise how do you know that this code does what you expect it to do (and will continue to do so in the future?).

File: src/main/webapp/loggedIn.html [hunk 6]
```diff
@@ -9,9 +9,8 @@
     <!-- Initialize firebase using CDN -->
     <script src="https://www.gstatic.com/firebasejs/ui/4.6.1/firebase-ui-auth.js"></script>
     <link type="text/css" rel="stylesheet" href="https://www.gstatic.com/firebasejs/ui/4.6.1/firebase-ui-auth.css" />
-    <script src="common.js"></script>
     <link href="style.css" rel="stylesheet" type="text/css" media="screen" />
-    <script src="app.js"></script>
+    
   </head>
   <body onload="addMenu()">
```
Comment (alexcepoi) [inline src/main/webapp/loggedIn.html:17]: It's generally good practice to put the onload callback in javascript rather than in html.
https://www.javascripttutorial.net/javascript-dom/javascript-onload/

Then you don't need to load the scripts after the body.

**Review by alexcepoi — CHANGES_REQUESTED** [review]

File: src/main/java/com/google/sps/servlets/UserDataServlet.java [hunk 7]
```diff
@@ -1,59 +1,57 @@
...
-  
+
     // Initialize firebase app.
-    FirebaseApp defaultApp = initializeFirebaseApp(/* setOptions */ false);
+    FirebaseApp firebaseApp = FirebaseSingletonApp.getInstance();
     String uid = null;
     try {
-      // get the idToken from hidden input field
+      // get the idToken from query string
       String idToken = request.getParameter("idToken");
```
Comment (ak47na) [inline src/main/java/com/google/sps/servlets/UserDataServlet.java]: I'm not sure if it's a problem with sending the token ID (a private information of the user) in a GET request, but I found that in the documentation, it's recommended to send it via POST: 
https://developers.google.com/identity/sign-in/android/backend-auth

File: src/main/java/com/google/sps/data/FirebaseSingletonApp.java (hunk 1 above)
Comment (anamariamilcu) [reply src/main/java/com/google/sps/data/FirebaseSingletonApp.java]: When I try to throw it in the signature of the initialize() function I get an error at 
private static FirebaseApp instance = initialize();
and I was not sure how else I can solve this.

File: src/main/java/com/google/sps/data/FirebaseSingletonApp.java (hunk 2 above)
Comment (anamariamilcu) [reply src/main/java/com/google/sps/data/FirebaseSingletonApp.java]: Done

File: src/main/java/com/google/sps/servlets/NewRequestServlet.java (hunk 4 above)
Comment (anamariamilcu) [reply src/main/java/com/google/sps/servlets/NewRequestServlet.java]: Done

File: src/main/java/com/google/sps/data/FirebaseAuthentication.java [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+  public FirebaseApp getFirebaseApp() {
+    return firebaseApp;
+  }
+  
+  public String getUserIdFromIdToken(String idToken) throws FirebaseAuthException {
+    FirebaseToken decodedToken = FirebaseAuth.getInstance(firebaseApp).verifyIdToken(idToken);
+    return decodedToken.getUid();
+  }
+
+  public UserRecord getUserFromUserId(String uid) throws FirebaseAuthException {
```
Comment (alexcepoi) [inline src/main/java/com/google/sps/data/FirebaseAuthentication.java]: This is not used. Do not polute your class with unused public methods and always aim to keep the number of public methods minimal. The fewer public things in your classes the better.

Same for getFirebaseApp.

File: src/main/java/com/google/sps/data/FirebaseSingletonApp.java [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+   */
+
+  private static FirebaseApp initialize() {
+    try {
+      FirebaseOptions options = new FirebaseOptions.Builder()
+          .setCredentials(GoogleCredentials.getApplicationDefault())
+          .setDatabaseUrl("https://alphabooks-step-2020.firebaseio.com").build();
+
+      return FirebaseApp.initializeApp(options, "AlphaBooks");
+    } catch (IOException e) {
```
Comment (alexcepoi) [inline src/main/java/com/google/sps/data/FirebaseSingletonApp.java]: Why? Do not catch exceptions you don't understand. Exceptions are good.

File: src/test/java/com/google/sps/servlets/NewRequestServletTest.java [hunk 10]
```diff
@@ -45,17 +59,32 @@
...
+
+  String idToken;
+  String userId;
+
   @Before
   public void setUp() {
+    helper.setUp();
     // initialize Mock objects
     MockitoAnnotations.initMocks(this);
+    idToken = "idToken0";
```
Comment (alexcepoi) [inline src/test/java/com/google/sps/servlets/NewRequestServletTest.java]: nit: No need for these constants here, just use the "idToken0", "user0" values directly in your test.

File: src/test/java/com/google/sps/servlets/NewRequestServletTest.java [hunk 11]
```diff
@@ -95,23 +124,33 @@ public void testStartTimeAfterEndTime() throws IOException, ServletException {
...
-//     NewRequestServlet requestServlet = new NewRequestServlet();
-//     requestServlet.doPost(request, response);
-//     // the date is valid, thus "Please enter a valid date!" shouldn't be sent
-//     verify(response, times(0)).sendError(HttpServletResponse.SC_BAD_REQUEST, "Please enter a valid date!");
-//     verify(response, times(0)).sendError(HttpServletResponse.SC_BAD_REQUEST, "Please enter a valid time!");
-//     verify(response, times(0)).sendError(HttpServletResponse.SC_BAD_REQUEST, "Please enter between 1 and 25 stops!");
-//     //TODO[ak47na]: check that the delivery request is added to Mock datastore
-//   }
+  @Test
+  public void testValidData() throws IOException, ServletException, FirebaseAuthException {
```
Comment (alexcepoi) [inline src/test/java/com/google/sps/servlets/NewRequestServletTest.java]: Test description and comment don't make much sense.

Maybe you meant something like testNewRequestsAutomaticallyCreatesCourierProfile?

File: src/main/webapp/common.js [hunk 12]
```diff
@@ -20,14 +20,23 @@
...
-    });
+  firebase.auth().onAuthStateChanged(function(user) {
+  if (user) {
+    // User is signed in.
+    if (document.getElementById('idToken')) {
+      user.getIdToken(/* forceRefresh */ true).then(function(idToken) {
+        document.getElementById('idToken').value = idToken;
+      });
+    }
+    else {
```
Comment (alexcepoi) [inline src/main/webapp/common.js]: nit: "} else {"

https://google.github.io/styleguide/jsguide.html#formatting-nonempty-blocks

File: src/main/webapp/common.js (hunk 5 above)
Comment (alexcepoi) [reply src/main/webapp/common.js:120]: We agreed to drop dom manipulation tests.

File: src/main/webapp/common.js [hunk 13]
```diff
@@ -72,9 +80,15 @@ function parseQueryString(queryString) {
  */
 
 function displayCurrentAddress() {
```
Comment (alexcepoi) [inline src/main/webapp/common.js:81]: Why is this functoin setting the onAuthStateChanged twice? Can we only do it once?

File: src/main/java/com/google/sps/servlets/UserDataServlet.java [hunk 14]
```diff
@@ -1,59 +1,65 @@
...
     try {
-      // get the idToken from hidden input field
-      String idToken = request.getParameter("idToken");
-      FirebaseToken decodedToken = FirebaseAuth.getInstance().verifyIdToken(idToken);
-      uid = decodedToken.getUid();
+      // get the idToken from query string
+      uid = firebaseAuth.getUserIdFromIdToken(request.getParameter("idToken"));
     } catch (FirebaseAuthException e) {
-      response.sendError(HttpServletResponse.SC_BAD_REQUEST, "Invalid idToken");
+      response.sendError(HttpServletResponse.SC_BAD_REQUEST, e.getMessage());
```
Comment (alexcepoi) [inline src/main/java/com/google/sps/servlets/UserDataServlet.java]: nit: SC_UNAUTHORIZED might work better here if the user is unauthenticated.

Same elsewhere.

https://docs.oracle.com/javaee/6/api/javax/servlet/http/HttpServletResponse.html

File: src/main/java/com/google/sps/servlets/UserDataServlet.java [hunk 15]
```diff
@@ -1,59 +1,65 @@
...
       return;
     }
 
-    /* Return all info about the current user if he has been added to the database. */
+    /*
```
Comment (alexcepoi) [inline src/main/java/com/google/sps/servlets/UserDataServlet.java:57]: Please stop fixing unrelated stuff in PRs :(

If it bothers you do a completely separate PR to fix the whitespace.

File: src/test/java/com/google/sps/servlets/NewRequestServletTest.java (hunk 11 above)
Comment (alexcepoi) [inline src/test/java/com/google/sps/servlets/NewRequestServletTest.java]: This looks ok for now.

For future reference I would advise to get everything out of the servlet code in order to make testing easier. This avoids having to mock a lot of the firebase stuff in the future:
https://github.com/googleinterns/step235-2020/pull/1/files#r481166522

File: src/test/java/com/google/sps/servlets/NewRequestServletTest.java [hunk 16]
```diff
@@ -95,23 +124,33 @@ public void testStartTimeAfterEndTime() throws IOException, ServletException {
...
+    
+    requestServlet.setFirebaseAuth(firebaseAuth);
+    requestServlet.doPost(request, response);
+
+    DatastoreService ds = DatastoreServiceFactory.getDatastoreService();
+    // The user is a new user and it is added to datastore.
+    Query userQuery = new Query("UserData")
+          .setFilter(new Query.FilterPredicate("uid", Query.FilterOperator.EQUAL, userId));
+    assertEquals(1, ds.prepare(userQuery).countEntities(FetchOptions.Builder.withLimit(10)));
+    // TODO[ak47na]: update test to check that the delivery request is added after
```
Comment (alexcepoi) [inline src/test/java/com/google/sps/servlets/NewRequestServletTest.java]: Test should be super focused and minimal.

You should not have a single test that tests "valid data". You should have multiple very specific ones, e.g.:

- testNewRequestsAutomaticallyCreatesCourierProfile
- testNewRequestEnqueuesNewDeliveryRequestInDataStore
- testNewRequestFailsForUnauthorizedUser

etc.

**Review by alexcepoi — APPROVED** [review]

File: src/main/java/com/google/sps/data/FirebaseAuthentication.java (hunk 8 above)
Comment (ak47na) [reply src/main/java/com/google/sps/data/FirebaseAuthentication.java]: I removed them.

File: src/main/java/com/google/sps/data/FirebaseSingletonApp.java (hunk 9 above)
Comment (ak47na) [reply src/main/java/com/google/sps/data/FirebaseSingletonApp.java]: I changed the code to catch IOException in the init() function of the servlets that use firebaseApp. I the next PR, I will restructure the code so that there is no need to catch the exception.

File: src/test/java/com/google/sps/servlets/NewRequestServletTest.java (hunk 10 above)
Comment (ak47na) [reply src/test/java/com/google/sps/servlets/NewRequestServletTest.java]: I removed the idToken, userId constants.

File: src/main/java/com/google/sps/servlets/UserDataServlet.java (hunk 14 above)
Comment (ak47na) [reply src/main/java/com/google/sps/servlets/UserDataServlet.java]: I changed the error message.

File: src/test/java/com/google/sps/servlets/NewRequestServletTest.java (hunk 11 above)
Comment (ak47na) [reply src/test/java/com/google/sps/servlets/NewRequestServletTest.java]: Yes, I will refactor the code in https://github.com/googleinterns/step235-2020/pull/1/files#r481166522.

File: src/test/java/com/google/sps/servlets/NewRequestServletTest.java (hunk 16 above)
Comment (ak47na) [reply src/test/java/com/google/sps/servlets/NewRequestServletTest.java]: I changed the name from testValidData() to testNewRequestsAutomaticallyCreatesCourierProfile() and I will add  

- testNewRequestEnqueuesNewDeliveryRequestInDataStore

- testNewRequestFailsForUnauthorizedUser

File: src/main/webapp/common.js (hunk 13 above)
Comment (anamariamilcu) [reply src/main/webapp/common.js:81]: The reason behind this is that we want to extract the idToken of the user and the idToken can modify during a session so we considered it's safer to get it every time we need it.
